packages/rum-core/src/domain/rumEventsCollection/action/clickChain.ts

```diff
@@ -0,0 +1,100 @@
+import type { TimeStamp } from '@datadog/browser-core'
+import { monitor, ONE_SECOND } from '@datadog/browser-core'
+
+export interface ClickChain<Click extends BaseClick> {
+  tryAppend: (click: Click) => null | ClickReference
+  stop: () => void
+}
+
+export interface ClickReference {
+  markAsComplete: () => void
+}
+
+export interface BaseClick {
+  event: MouseEvent
+  timeStamp: TimeStamp
+}
+
+export const CLICK_CHAIN_WINDOW_SIZE = 3
+export const CLICK_CHAIN_MAX_DURATION_WINDOW = ONE_SECOND
+export const CLICK_CHAIN_MAX_DISTANCE_WINDOW = 100
+
+const enum ClickChainStatus {
+  WaitingForMoreClicks,
+  WaitingForClicksToComplete,
+  Flushed,
+}
+
+export function createClickChain<Click extends BaseClick>(onFlush: (clicks: Click[]) => void): ClickChain<Click> {
+  const clicks: Click[] = []
+  let completeClicksCount = 0
+  let status = ClickChainStatus.WaitingForMoreClicks
+  let timeout: number | undefined
+
+  function dontAcceptMoreClick() {
+    if (status === ClickChainStatus.WaitingForMoreClicks) {
+      status = ClickChainStatus.WaitingForClicksToComplete
+      tryFlush()
+    }
+  }
+
+  function tryFlush() {
+    if (status === ClickChainStatus.WaitingForClicksToComplete && completeClicksCount === clicks.length) {
+      status = ClickChainStatus.Flushed
+      onFlush(clicks)
+    }
+  }
+
+  function getWindowStartClick() {
+    return clicks[Math.max(0, clicks.length - CLICK_CHAIN_WINDOW_SIZE)]
+  }
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** that seems pretty odd to me, what is the reasoning for doing that?‏